bump github.com/coreos/go-oidc dep to pick up nbf enforcement

anderseknert opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
If the nbf (not before) claim is present in an ID token, check that it's indeed in the past. This allows the token issuer to control the point in time at when the token should be usable for authentication.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
The nbf (not before) claim, if present in ID token, is now enforced.  
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

anderseknert committed:
Honour "nbf" claim if present in ID token

ericchiang commented:
/ok-to-test

ericchiang commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:
```diff
@@ -633,6 +636,21 @@ func (a *Authenticator) AuthenticateToken(ctx context.Context, token string) (*a
+		}
+
+		nbfTime := time.Unix(nbfValue, 0)
+		now := a.now()
```
now() can be nil here. This needs to default to time.Now above.

anderseknert replied:
Ah, alright - golang new to me so still trying to figure out all the conditions for when nil checks are needed. I thought the check at line 308 would be enough, but coming to think of it of course that only covers the case where the authenticator is constructed by the `newAuthenticator` function. Any good tools for finding possible nil-problems like there is for Java? I'll fix.

ericchiang replied:
I don't know of a common one. Generally, you ensure this during initialization:

```
now := config.now
if now == nil {
    now = time.Now
}
a := authenticator{
    now: now,
    // ...
}
```

mikedanese replied:
Or  `a := authenticator{now: time.Now}` and set over it in tests.

ericchiang commented:
This might be a better feature request for github.com/coreos/go-oidc, since that's what does the current expiration validation.

anderseknert commented:
I considered that too, but `nbf´ doesn't make much sense outside of authentication contexts IMO - and as that library is more generic I thought it would fit better here. If you think I should try submitting a PR there first I'll do that.

ericchiang commented:
I think `nbf` makes sense for ID Token validation in go-oidc instead of having it only in Kubernetes. Want to submit it there, then we can pull it into kube?

cc @mikedanese

mikedanese reviewed (commented):
Adding this to go-oidc makes sense to me as well. Can you send a change to https://github.com/coreos/go-oidc, we'll review there, and then we'll bump the dep?

anderseknert commented:
Thanks guys! PR targeting go-oidc coming up 👍

fedebongio commented:
/remove-sig api-machinery

mikedanese commented:
@anderseknert do you want to bump the go-oidc dependency?

anderseknert commented:
I'd be delighted @mikedanese . Just heading to bed here but will check it out first thing tomorrow after having read up on how that works, hah. It's been three weeks since I wrote my first line of Go so it's all been moving pretty fast. Any pointers would be appreciated :)

mikedanese commented:
updating deps in kubernetes is no easy task. will give it a shot and report instructions.

mikedanese commented:
/lgtm
/hold

This needs a release note. Can you update the description?

anderseknert commented:
@mikedanese ✅

liggitt commented on go.mod:
```diff
@@ -219,7 +219,7 @@ replace (
 	github.com/coreos/etcd => github.com/coreos/etcd v3.3.13+incompatible
 	github.com/coreos/go-etcd => github.com/coreos/go-etcd v2.0.0+incompatible
-	github.com/coreos/go-oidc => github.com/coreos/go-oidc v0.0.0-20180117170138-065b426bd416
+	github.com/coreos/go-oidc => github.com/coreos/go-oidc v2.0.1-0.20190815175729-2be1c5b8a260+incompatible
```
I won't stick on it since we weren't on a tag before, but any chance of getting a v2.1.0 tag cut for this? cc @ericchiang

ericchiang replied:
Yeah. I'll do that today

ericchiang replied:
Tagged https://github.com/coreos/go-oidc/releases/tag/v2.1.0 :)

liggitt commented:
> Edit: short answer is it's really hard. Maybe there are some docs I don't know about.

See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/vendor.md#adding-or-updating-a-dependency

anderseknert commented:
/retest

anderseknert commented:
That's a great resource, thanks @liggitt! Will try to do it myself next time. About the failing tests reported, I don't see how they are related to any changes made here. Are they known to be flaky, or something else that changed? I'm on a 2 week leave so a little hard trying to follow when not on my computer. Again - thanks for all the help guys!

liggitt commented:
/lgtm
/approve

liggitt commented:
unfortunately, yes, sometimes the scale tests flake

anderseknert commented:
Anything more needed from me here?

anderseknert commented:
@liggitt I've now rebased, fixed conflicts and run `hack/update-vendor.sh` which checked out (thanks for the pointer, btw). As the conflicting changes are in go.mod|sum though I expect this will diverge from master soon again though.

liggitt commented:
Thanks!

/lgtm
